File: _C_satellite/simulink/satellite_animation.m
function satellite_animation(u, P)

    % process inputs to function
    theta    = u(1);
    phi      = u(2);
    %thetadot = u(3);
    %phidot   = u(4);
    t        = u(5);
    
    % define persistent variables 
    persistent base_handle
    persistent panel_handle
    
    % initialize plot and persistent vars at t=0
    if t==0
        figure(1), clf
        % plot track
        plot([-2*P.length,2*P.length],[0,0],...
            'k--'); 
        hold on
        base_handle...
            = drawBase(theta, P.width, []);
        panel_handle...
            = drawPanel(phi,P.width,P.length,[]);
        axis([-2*P.length, 2*P.length,...
              -2*P.length, 2*P.length]);
    
        
    % otherwise, redraw base and rod
    else 
        drawBase(theta, P.width, base_handle);
        drawPanel(phi, P.width, P.length, ...
                  panel_handle);
    end
end

   
%
%=============================================
% drawBase
% draw the base of the pendulum
% return handle if 3rd argument is empty, 
% otherwise use 3rd arg as handle
%=============================================
%
function handle = drawBase(theta, w, handle)

  % define points on base (without rotation)
  pts = [...
      w/2, -w/2;...
      w/2, -w/6;...
      w/2+w/6, -w/6;...
      w/2+w/6,  w/6;...
      w/2, w/6;...
      w/2, w/2;...
      -w/2, w/2;...
      -w/2, w/6;...
      -w/2-w/6, w/6;...
      -w/2-w/6, -w/6;...
      -w/2, -w/6;...
      -w/2, -w/2;...
      ]';
  % define rotation matrix
  R = [cos(theta), sin(theta);...
       -sin(theta), cos(theta)];
  % rotate points
  pts = R*pts;
  % break into X and Y components
  X = pts(1,:);
  Y = pts(2,:);

  if isempty(handle)
    handle = fill(X,Y,'b');
  else
    set(handle,'XData',X,'YData',Y);
    drawnow
  end
end
 
%
%=============================================
% drawPanel
% draw the solar panel
% return handle if 3rd argument is empty, 
% otherwise use 3rd arg as handle
%=============================================
%
function handle = drawPanel(phi, w, L, handle)

% define points on base (without rotation)
  pts = [...
      -L, -w/6;...
       L, -w/6;...
       L,  w/6;...
      -L,  w/6;...
      ]';
  % define rotation matrix
  R = [cos(phi), sin(phi);...
      -sin(phi), cos(phi)];
  % rotate points
  pts = R*pts;
  % break into X and Y components
  X = pts(1,:);
  Y = pts(2,:);

  if isempty(handle)
    handle = fill(X, Y, 'g');
  else
    set(handle,'XData',X,'YData',Y);
    drawnow
  end
end

  

File: _C_satellite/python/hw2/satelliteAnimation.py
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np 
import satelliteParam as P


class satelliteAnimation:
    '''
        Create satellite animation
    '''
    def __init__(self):
        # Used to indicate initialization
        self.flagInit = True
        
        # Initializes a figure and axes object
        self.fig, self.ax = plt.subplots()

        # Initializes a list object that will be used to contain
        # handles to the patches and line objects.
        self.handle = []
        
        plt.axis([-2.0*P.length, 2.0*P.length, -2.0*P.length,
                  2.0*P.length])
        plt.plot([-2.0*P.length, 2.0*P.length], [0, 0], 'b--')
        self.length = P.length
        self.width = P.width

    def update(self, u):
        # Process inputs to function
        theta = u.item(0)   # Angle of base, rad
        phi = u.item(1)     # angle of panel, rad

        self.drawBase(theta)
        self.drawPanel(phi)

        # This will cause the image to not distort
        # self.ax.axis('equal') 

        # After each function has been called, initialization is
        # over.
        if self.flagInit == True:
            self.flagInit = False

    def drawBase(self, theta):
        # points that define the base
        pts =np.matrix([
            [self.width/2.0, -self.width/2.0],
            [self.width/2.0, -self.width/6.0],
            [self.width/2.0 + self.width/6.0, -self.width/6.0],
            [self.width/2.0 + self.width/6.0, self.width/6.0],
            [self.width/2.0, self.width/6.0],
            [self.width/2.0, self.width/2.0],
            [-self.width/2.0, self.width/2.0],
            [-self.width/2.0, self.width/6.0],
            [-self.width/2.0 - self.width/6.0, self.width/6.0],
            [- self.width/2.0 - self.width/6.0, -self.width/6.0],
            [- self.width/2.0, -self.width/6.0],
            [-self.width/2.0, -self.width/2.0]]).T
        R = np.matrix([[np.cos(theta), np.sin(theta)],
                       [-np.sin(theta), np.cos(theta)]])
        pts = R*pts
        xy = np.array(pts.T)

        # When the class is initialized, a polygon patch object
        # will be created and added to the axes. After
        # initialization, the polygon patch object will only be
        # updated.
        if self.flagInit == True:

            # Create the Rectangle patch and append its handle
            # to the handle list
            self.handle.append(mpatches.Polygon(xy,
                                                facecolor='blue',
                                                edgecolor='black'))
            # Add the patch to the axes
            self.ax.add_patch(self.handle[0]) 

        else:

            # Update polygon
            self.handle[0].set_xy(xy)         

    def drawPanel(self, phi):
            # points that define the base
            pts = np.matrix([
                [-self.length, -self.width/6.0],
                [self.length, -self.width/6.0],
                [self.length, self.width/6.0],
                [-self.length, self.width/6.0]]).T
            R = np.matrix([[np.cos(phi), np.sin(phi)],
                           [-np.sin(phi), np.cos(phi)]])
            pts = R * pts
            xy = np.array(pts.T)

            # When the class is initialized, a polygon patch
            # object will be created and added to the
            # axes. After initialization, the polygon patch
            # object will only be updated.
            if self.flagInit == True:

                # Create the Rectangle patch and append its
                # handle to the handle list
                self.handle.append(mpatches.Polygon(xy,
                                   facecolor='green',
                                   edgecolor='black'))

                # Add the patch to the axes
                self.ax.add_patch(self.handle[1])  

            else:

                # Update polygon
                self.handle[1].set_xy(xy)  
